hypershift-operator/controllers/hostedcluster/internal/platform/powervs/ccm.go

```diff
@@ -0,0 +1,302 @@
+package powervs
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"strings"
+	"text/template"
+
+	ibmcore "github.com/IBM/go-sdk-core/v5/core"
+	"github.com/IBM/platform-services-go-sdk/iamidentityv1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/upsert"
+)
+
+const (
+	// ConfigMap
+	configMapName         = "cloud-conf"
+	kubeConfigKeyInSecret = "kubeconfig"
+
+	// Deployment
+	deploymentName         = "ibm-cloud-controller-manager"
+	containerName          = "cloud-controller-manager"
+	kubeConfigFileBasePath = "/etc/kubernetes"
+	secretMountPath        = "/etc/vpc"
+	configMapMountPath     = "/etc/ibm"
+)
+
+type configMap struct {
+	ConfigFile               string
+	AccountID                string
+	ClusterID                string
+	G2workerServiceAccountID string
+	G2Credentials            string
+	G2ResourceGroupName      string
+	G2VpcSubnetNames         string
+	G2VpcName                string
+	Region                   string
+	PowerVSCloudInstanceID   string
+	PowerVSRegion            string
+	PowerVSZone              string
+}
+
+type setupCCM struct {
+	ctx            context.Context
+	c              client.Client
+	hcluster       *hyperv1.HostedCluster
+	createOrUpdate upsert.CreateOrUpdateFN
+	kubeConfig     *corev1.Secret
+	image          string
+}
+
+func (ccm setupCCM) getAccountID() (accountID string, err error) {
+	var secret corev1.Secret
+	err = ccm.c.Get(ccm.ctx, client.ObjectKey{Namespace: ccm.hcluster.GetNamespace(), Name: ccm.hcluster.Spec.Platform.PowerVS.ClusterControlManagerCreds.Name}, &secret)
+	if err != nil {
+		err = fmt.Errorf("failed to get control plane operator provider creds %s: %w", ccm.hcluster.Spec.Platform.PowerVS.ClusterControlManagerCreds.Name, err)
+		return
+	}
+	secretData, secretHasData := secret.Data["ibmcloud_api_key"]
+	if !secretHasData {
+		err = fmt.Errorf("cluster control manager credentials secret %q is missing credentials key", secret.Name)
+		return
+	}
+	apiKeyStr := string(secretData)
+	auth := &ibmcore.IamAuthenticator{
+		ApiKey: apiKeyStr,
+	}
+
+	iamv1, err := iamidentityv1.NewIamIdentityV1(&iamidentityv1.IamIdentityV1Options{
```

Comment:
What is the reason we have to do this api call each and every time we reconcile rather than asking the user to submit it or maybe default and persist it?